### PR TITLE
fix_rollback_conflict_for_vol_attach_failure

### PR DESCRIPTION
**Pull request by jackydalong**
### Operations across 3 Layers are involved in volume attach-or-detachment
1. Dedicate/undedicate FCP devices in z/VM layer
2. Online/offline volume in virtual machine operating system(OS)
3. Allocate/release FCP devices in sqlite database
 
### The PR fixes the following items:
1. fix the rollback conflict for volume attachment
4. fix the rollback conflict for volume detachment
5. enhance log for rollback of attach/detach

Signed-off-by: Da Long Wang <shdlwang@cn.ibm.com>

### CODE REVIEW

**Comment by jackydalong** [comment]
### fix the rollback conflict for volume Attachment

- background:
  - By design, we must ensure the rollback logic in 'API attach' can
    - rollback the action done 
    - by both 'API attach' and 'API get_volume_connector'.
  - image

- there is a conflict between 2 rollbacks for connections
  - image

- The following alternative proposal is NOT thread safe
  - image

- the PR fix it as follows
  - image

**Comment by jackydalong** [comment]
### fix the rollback conflict for volume Detachment

- there is a conflict between 2 rollbacks for connections
  - image

- the PR fix it as follows
  - image

**Comment by jackydalong** [comment]
I'm writing the UT

File: zvmsdk/volumeop.py [hunk 1]
```diff
@@ -1907,30 +1960,14 @@ def _detach(self, fcp_list, assigner_id, target_wwpns, target_lun,
                 multipath, os_version, mount_point, is_root_volume,
                 update_connections_only):
         """Detach a volume from a guest"""
-        LOG.info("Start to detach volume on machine %s from "
+        LOG.info("Start to detach volume on virtual machine %s from "
                  "FCP devices %s" % (assigner_id, fcp_list))
         with database.get_fcp_conn():
```
Comment (IBMCaoBiao) [inline zvmsdk/volumeop.py:1965]: do we need `with database._DBLOCK_FCP:` in `_detach` ?

**Review by IBMCaoBiao — APPROVED** [review]
I can not come up with error cases that current patch do NOT apply to.
It is LGTM now. we must add as more UT case as possible to make sure it works as expected.

File: zvmsdk/volumeop.py (hunk 1 above)
Comment (jackydalong) [reply zvmsdk/volumeop.py:1965]: in `_detach()`, `_DBLOCK_FCP` is acquired implicitly at the beginning by `with database.get_fcp_conn()`

File: zvmsdk/volumeop.py [hunk 2]
```diff
@@ -1755,26 +1754,38 @@ def _add_disks(self, fcp_list, assigner_id, target_wwpns, target_lun,
...
+            # undedicate FCP device from assigner_id
             for fcp in fcp_list:
                 with zvmutils.ignore_errors():
-                    LOG.info("Rolling back dedicated FCP: %s" % fcp)
-                    connections = self.fcp_mgr.decrease_fcp_usage(
-                        fcp, assigner_id)
+                    connections = self.db.get_connections_from_fcp(fcp)
                     if connections == 0:
+                        no_connection_fcps.append(fcp)
                         self._undedicate_fcp(fcp, assigner_id)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1779]: for this undedicate, I think we should ignore errors like "FCP device not exist", right?

**Review by dongyanyang — APPROVED** [review]

File: zvmsdk/volumeop.py (hunk 2 above)
Comment (jackydalong) [reply zvmsdk/volumeop.py:1779]: The `with zvmutils.ignore_errors()` ignore any Exception, 
so as to ensure _undedicate_fcp() to be run on the last FCP device in fcp_list,
even if _undedicate_fcp() raise any exception on the previous FCP devices in fcp_list

File: zvmsdk/volumeop.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackydalong) [reply zvmsdk/volumeop.py:1965]: - all the code of _detach() is included in the `with database.get_fcp_conn():`, 
  - `get_fcp_conn()` acquires _DBLOCK_FCP implicitly.
- So _detach() and _attach() are both thread safe

**Pull request merged**